### ISSUE TITLE
How to index a tree that stores elements at each node?

### ISSUE THREAD

**Issue by Gabriel439** [ekmett/succinct#9]
Let's assume that I have something like `Tree` from `Data.Tree` in `containers` and I want to efficiently navigate this tree using a zipper into a compressed index to retrieve elements from the tree.  My rough understanding is that the way I would do this is:

1.  Define my own function from `Tree` to `Succinct.Tree.LOUDS.Rose` that just extracts the tree's skeleton

2.  Compute a `Rank9` index from my `Rose`-encoded skeleton using `Succinct.Tree.LOUDS.fromRose`

3. Store the index alongside an array containing the tree elements in some order.  I'm guessing they would be stored in level order (for LOUDS), but I'm not totally sure.

4. When it's time to retrieve an element, wrap the `Rank9` index in a `Zipper` using `root` so I can navigate the skeleton of the tree without materializing it using all the zipper operations

5. After I navigate to some node, I need to retrieve the index into my array of elements.  This is the part I don't know how to do since I'm not sure how I retrieve that index from the `Zipper`.  I read the documentation for `Zipper` but it's still pretty unclear to me if one of the two `Int` fields is the index or not.

So basically I need help with steps 3 and 5: what order do I store the array of elements and how do I retrieve the index into the array from the `Zipper`?

**Comment by klao** [comment]
For 3: yes the elements should be in level order. (And, it should be possible to extract the LOUDS structure and the vector of elements from a `Tree` in one go, if that's important.)

For 5: it's the `index` of the zipper. :) Or more precisely `index z - 1`, because it's numbered from 1 (`index . root == const 1`).

And now to the bad news. The thing is unfinished; in particular, `select` is not implemented for `Rank9`. (Well, it is, but only in the generic binary search way, not in constant time as it should be.)  So, as this currently is, the navigation wouldn't be constant time.

I should probably go back and implement select for Poppy and Rank9. (I got stalled, as always, not with the coding, but naming and organizing stuff. Should we always build the auxiliary structures both for Select0 and Select1? But, what if someone only needs one of them...)

**Comment by Gabriel439** [comment]
Oops, I totally missed the `index` function for some reason :)

I'm okay with `select` being `O(log n)` for now.  I just wanted to play with this a little bit.
